### PR TITLE
CodeAction to fix unknown property by proposing similar name

### LINKED ISSUE #80: CodeAction to fix unknown property by proposing similar name
Given an unknown property:

`quarkus.application.nme=`

A code action could proposed to fix it with

`Did you mean with quarkus.application.name ?` 

To do that @fbricon suggest to use 

https://commons.apache.org/proper/commons-text/apidocs/org/apache/commons/text/similarity/LevenshteinDistance.html

Here a demo:

QuarkusCodeActionForUnknownDemo

### PR DESCRIPTION
**Pull request by angelozerr**
CodeAction to fix unknown property by proposing similar name

Fixes #80

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
Here a demo:

QuarkusCodeActionForUnknownDemo

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusTextDocumentService.java [hunk 1]
```diff
@@ -218,6 +226,37 @@ public void didSave(DidSaveTextDocumentParams params) {
...
+								sharedSettings.getFormattingSettings()) //
+						.stream() //
+						.map(ca -> {
+							if (codeActionLiteralSupport) {
+								Either<Command, CodeAction> e = Either.forRight(ca);
+								return e;
+							} else {
+								List<Object> arguments = Arrays.asList(uri, document.getDocument().getVersion(),
+										ca.getEdit().getDocumentChanges().get(0).getLeft().getEdits());
+								Command command = new Command(ca.getTitle(), "_quarkus.applyCodeAction", arguments);
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusTextDocumentService.java]: do we really still need client specific implementations to apply code actions?

File: quarkus.ls/com.redhat.quarkus.ls/pom.xml [hunk 2]
```diff
@@ -111,6 +111,11 @@
 			<artifactId>org.eclipse.lsp4j.jsonrpc</artifactId>
 			<version>${lsp4j.version}</version>
 		</dependency>
+		<dependency>
+		    <groupId>org.apache.commons</groupId>
+		    <artifactId>commons-text</artifactId>
+		    <version>1.8</version>
+		</dependency>
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/pom.xml]: mmm we're adding 700KB (200KB of commons-text + 500KB of commons-lang) just for that feature. Seems overweight.
We should try to use proper shading to remove unnecessary cruft.

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+					codeActions.add(replaceAction);
+				}
+			}
+		} catch (BadLocationException e) {
+			LOGGER.log(Level.SEVERE, "In QuarkusCodeActions, position error", e);
+		}
+	}
+
+	private static boolean isSimilar(String left, String right) {
+		int threshold = Math.round(MAX_DISTANCE_DIFF_RATIO * left.length());
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java]: does it make sense to determine the threshold as 10% of the length of a word?

Alos it should be the distance from the target value, which is `right` here.

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusTextDocumentService.java (hunk 1 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusTextDocumentService.java]: It's a copy paste from LSP4XML code action. If you wish I can remove it.

File: quarkus.ls/com.redhat.quarkus.ls/pom.xml (hunk 2 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/pom.xml]: LevenshteinDistance from commons-text has none dependencies (just implement an interface but we can skip it), so if we can copy/paste this class, what do you think?

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java (hunk 3 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java]: It should be fixed.

File: quarkus.ls/com.redhat.quarkus.ls/pom.xml (hunk 2 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/pom.xml]: LevenshteinDistance class have none dependency (just one for implementing EditDistance interface) so I think we can copy/paste this class.

Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/pom.xml]: @fbricon I do a copy/paste of LevenshteinDistance  to remove the commons-text dependency.

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusTextDocumentService.java (hunk 1 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusTextDocumentService.java]: @fbricon do you want i remove this code?

**Comment by xorye** [comment]
Great PR

Should we support mapped properties too?
Gif showing codeaction with mapped properties

When the key is long, it is not within the threshold, and we get no code action even though I only removed "d" from "id".

When we do get a code action for the mapped property, the code action inserts the "{*}".

**Comment by angelozerr** [comment]
> Should we support mapped properties too?

Indeed I have not supported it. Let me see if it's easy to support it otherwise we could fix that in an another PR if @fbricon is agree with that

**Comment by angelozerr** [comment]
@xorye we decide to not support mapped property, please review my code again (and javadoc which speak about this LIMITATION), thanks!

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	 * @param projectInfo the Quarkus properties
+	 * @param codeActions code actions list to fill.
+	 */
+	private void doCodeActionsForUnknown(Diagnostic diagnostic, PropertiesModel document,
+			QuarkusProjectInfo projectInfo, List<CodeAction> codeActions) {
+		try {
+			// Get property name by using the diagnostic range
+			PropertyKey propertyKey = (PropertyKey) document.findNodeAt(diagnostic.getRange().getStart());
+			String propertyName = propertyKey.getPropertyName();
+			// Loop for each metata property
```
Comment (xorye) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java]: "metata property" should be "metadata property"

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+			for (ExtendedConfigDescriptionBuildItem metaProperty : projectInfo.getProperties()) {
+				if (QuarkusPropertiesUtils.isMappedProperty(metaProperty.getPropertyName())) {
+					// FIXME: support mapped property
```
Comment (xorye) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java]: Do you think we might support mapped property in the future?

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java (hunk 4 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java]: @xorye fixed

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java (hunk 5 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java]: It should be fine but today I have not found a clean solution to do that.

**Comment by angelozerr** [comment]
@fbricon @xorye is there something you wish I do?

**Comment by xorye** [comment]
Looks good to me

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusTextDocumentService.java (hunk 1 above)
Comment (fbricon) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusTextDocumentService.java]: yup

**Pull request merged**